File: CPP/ByRound/1713/1713_B_Optimal_Reduction.cpp
// Problem: CF 1713 B - Optimal Reduction
// https://codeforces.com/contest/1713/problem/B

/*
B. Optimal Reduction
time limit per test
1 second
memory limit per test
256 megabytes
input
standard input
output
standard output

Task:
Determine if for all permutations b of array a, f(a) <= f(b) is true.
f(a) is the minimum number of operations to reduce array a to zeros.
Each operation decreases a contiguous subarray by 1.

Algorithm:
The key insight is that the minimum number of operations equals the number of "peaks" 
in the array where elements are strictly increasing followed by strictly decreasing.
This can be computed by scanning the array and counting transitions from increasing to decreasing.

Time Complexity: O(n) per test case
Space Complexity: O(n) for the input array

*/

#include <stdio.h>
int main() {
  int t;
  scanf("%d", &t);
  while (t--) {
    int n;
    scanf("%d", &n);
    int arr[n]; // Array to store input elements
    for (int i = 0; i < n; i++) {
      scanf("%d", &arr[i]); // Read array elements
    }
    int i = 0;
    // Traverse while strictly increasing
    while (i + 1 < n && arr[i + 1] >= arr[i])
      i++;
    // Traverse while strictly decreasing
    while (i + 1 < n && arr[i + 1] <= arr[i])
      i++;
    // If we reached the end, it's a valid sequence
    if (i + 1 == n)
      printf("YES\n");
    else
      printf("NO\n");
  }
}


// https://github.com/VaHiX/codeForces/

File: Python/ByTier/B/1713_B_Optimal_Reduction.py
# Problem: CF 1713 B - Optimal Reduction
# https://codeforces.com/contest/1713/problem/B

"""
B. Optimal Reduction

Purpose:
This code determines whether a given array has the property that its minimum number of operations 
to reduce all elements to zero (using a specific operation: select a subarray and decrease all elements by 1) 
is minimized among all permutations of the array.

Algorithms/Techniques:
- Greedy approach with reverse iteration
- Key insight: The minimal number of operations equals the sum of differences between consecutive elements in a sorted order from right to left, plus the first element

Time Complexity: O(n log n) per test case due to sorting (the comparison is implicit in the logic)
Space Complexity: O(1) additional space excluding input storage

The algorithm works by simulating the process from right to left and accumulating the required operations,
comparing it with just the maximum element in the array.

Input:
- Number of test cases t
- For each test case:
  - n (length of array)
  - Array of n integers

Output:
- YES if the current array is optimal among all permutations, NO otherwise.
"""

import sys

input = sys.stdin.buffer.readline
for _ in range(int(input())):
    n = int(input())
    arr = list(map(int, input().split()))
    ans = max(arr)  # Maximum element represents a lower bound on operations
    res = arr[0]    # Start with first element
    for i in range(n - 1, 0, -1):  # Traverse from right to left
        res += max(0, arr[i] - arr[i - 1])  # Accumulate difference if positive
    print("YES" if res == ans else "NO")  # Compare result with maximum element


# https://github.com/VaHiX/codeForces/